src/backend/distributed/executor/executor.c

```diff
@@ -995,21 +1005,136 @@ FindOrCreateWorkerSession(WorkerPool *workerPool, MultiConnection *connection)
 
 
 /*
- * RunDistributedExecution runs a distributed execution to completion. It
- * creates a wait event set to listen for events on any of the connections
- * and runs the connection state machine when a connection has an event.
+ * ShouldRunTasksSequentially returns true if each of the individual tasks
+ * should be executed one by one. Note that this is different than
+ * MultiShardConnectionType == SEQUENTIAL_CONNECTION case. In that case,
+ * running the tasks across the nodes in parallel is acceptable and implemented
+ * in that way.
+ *
+ * However, the executions that are qualified here would perform poorly if the
+ * tasks across the workers are executed in parallel. We currently qualify only
+ * one class of distributed queries here, multi-row INSERTs. If we do not enforce
+ * true sequential execution, concurrent multi-row upserts could easily form
+ * a distributed deadlock when the upserts touch the same rows.
+ */
+static bool
+ShouldRunTasksSequentially(DistributedExecution *execution)
+{
+	DistributedPlan *plan = execution->plan;
+	Job *workerJob = plan->workerJob;
+	List *taskList = workerJob->taskList;
+	Task *initialTask = NULL;
+
+	if (list_length(taskList) < 2)
+	{
+		/* single task plans are already qualified as sequential by definition */
+		return false;
+	}
+
+	/* all the tasks are the same, so we only look one */
+	initialTask = (Task *) linitial(taskList);
+	if (initialTask->rowValuesLists != NIL)
+	{
+		/*
+		 * We explicitly assert the fields to be NULL because we implemented the
+		 * single task plan executions for a restricted set of operations, e.g.,
+		 * only multi-row inserts. So, we don't expect other plans go through this
+		 * code, unless execution is implemented properly.
+		 */
+		Assert (plan->operation == CMD_INSERT);
+		Assert (plan->subPlanList == NIL);
+		Assert (plan->insertSelectSubquery == NULL);
+		Assert (plan->insertTargetList == NIL);
+		Assert (plan->intermediateResultIdPrefix == NULL);
+		Assert (plan->masterQuery == NULL);
+
+		/* found a multi-row INSERT */
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * SequentialRunDistributedExecution gets a distributed execution and
+ * executes each individual task in the exection sequentially, one
+ * task at a time. See related function ShouldRunTasksSequentially()
+ * for more detail on the definition of SequentialRun.
+ */
+static void
+SequentialRunDistributedExecution(DistributedExecution *execution)
+{
+	DistributedPlan *distributedPlan = execution->plan;
+	Job *job = distributedPlan->workerJob;
+	List *taskList = job->taskList;
+
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		Task *taskToExecute = (Task *) lfirst(taskCell);
+		DistributedPlan *singleTaskDistributedPlan = copyObject(distributedPlan);
```

Comment:
This is quite expensive to do on every iteration.

Reply:
I guess that's mostly because the tasks could be giant in this case, right? I'll push a commit which does only copy the required parts of the plan for multi-row inserts.